src/check/Check.zig

```diff
@@ -3621,21 +3621,85 @@ fn checkBinopExpr(
     does_fx = try self.checkExpr(binop.rhs, env, .no_expectation) or does_fx;
 
     switch (binop.op) {
-        .add => {
-            // For builtin numeric types, use the efficient special-cased numeric constraint logic
-            // For user-defined nominal types, desugar `a + b` to `a.plus(b)` using static dispatch
+        // Operators that support desugaring to method calls on nominal types
+        .add, .sub, .mul, .div, .div_trunc, .rem => {
+            // Map operator to its corresponding method name
+            const method_name: Ident.Idx = switch (binop.op) {
+                .add => self.cir.plus_ident,
+                .sub => self.cir.minus_ident,
+                .mul => self.cir.times_ident,
+                .div => self.cir.div_ident,
+                .div_trunc => self.cir.div_trunc_ident,
+                .rem => self.cir.rem_ident,
+                else => unreachable,
+            };
+
+            // Unwrap aliases to check the underlying type
+            var check_var = lhs_var;
+            var lhs_resolved = self.types.resolveVar(check_var);
+            while (lhs_resolved.desc.content == .alias) {
+                const alias_data = lhs_resolved.desc.content.alias;
+                check_var = self.types.getAliasBackingVar(alias_data);
+                lhs_resolved = self.types.resolveVar(check_var);
+            }
+
+            const lhs_content = lhs_resolved.desc.content;
 
-            // Check if lhs is a nominal type
-            const lhs_resolved = self.types.resolveVar(lhs_var).desc.content;
-            const is_nominal = switch (lhs_resolved) {
+            // Check if this is a known builtin number type (optimization)
+            const is_known_number = switch (lhs_content) {
+                .structure => |s| s == .num,
+                else => false,
+            };
+
+            // Check if we should use static dispatch (nominal types, or flex/rigid with constraints)
+            const should_use_static_dispatch = switch (lhs_content) {
                 .structure => |s| s == .nominal_type,
+                .flex => |f| f.constraints.len() > 0, // Flex with existing constraints
+                .rigid => |r| r.constraints.len() > 0, // Rigid with constraints from where clause
                 else => false,
             };
```

Comment:
i think all flex/rigids, regardless of existing constraints, need to go into the static dispatch codepath. if any non-constrained flex/rigids fallback to the `Num` inference, then this:
```
|a, b| a + b
```
would be inferred as `Num(size)`

when it should be inferred as:
```
a where [ a.plus : a -> b -> c ]
```